File: app/src/main/java/com/example/toutiao/Base/BaseActivity.java
package com.example.toutiao.Base;

import android.content.Context;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

public abstract class BaseActivity<P extends IBasePresenter> extends AppCompatActivity implements IBaseView {

    protected P mPresenter;

    protected abstract void initViews();

    protected abstract void initLayout(@Nullable Bundle savedInstanceState);

    protected abstract void initData();

    protected abstract P setPresenter();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        initLayout(savedInstanceState);

        mPresenter = setPresenter();
        mPresenter.attach(this);

        initViews();
        initData();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        /**
         * 解绑，以防止内存泄漏
         */
        mPresenter.detach();
        mPresenter = null;
    }

    @Override
    public Context getContext() {
        return this;
    }
}

File: app/src/main/java/com/example/toutiao/base/BaseActivity.kt
package com.example.toutiao.base

import android.app.Activity
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import com.blankj.utilcode.utils.BarUtils
import com.example.toutiao.R
import com.example.toutiao.utils.ActivityCollector
import permissions.dispatcher.RuntimePermissions
import java.lang.ref.WeakReference

@RuntimePermissions
open class BaseActivity : AppCompatActivity() {

    protected lateinit var activity: Activity

    protected lateinit var activityWR: WeakReference<Activity>

    protected var isActive = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        activity = this
        activityWR = WeakReference(this)
        ActivityCollector.addToStack(activityWR)
    }

    override fun setContentView(layoutResID: Int) {
        super.setContentView(layoutResID)

        BarUtils.setColor(this, R.color.colorPrimaryDark)
        setUpViews()
    }

    protected open fun setUpViews() {
    }

    override fun onResume() {
        super.onResume()
        isActive = true
    }

    override fun onPause() {
        super.onPause()
        isActive = false
    }

    override fun onDestroy() {
        super.onDestroy()
        ActivityCollector.removeFromStack(activityWR)
    }
}
